### ISSUE TITLE
Multi-GPU performance issue

### ISSUE THREAD

**Issue by mahbubnoor** [espnet/espnet#2761]
Hi everyone, I am getting lower CER score if i use multi-gpu with increasing batch size. When i executed given recipe of an4 i got CER like in the following if i use 1 gpu:

```
# CER results in exp/train_nodev_pytorch_train_mtlalpha1.0/decode_test_decode_ctcweight1.0_lm_word100/result.txt
|  SPKR     |  # Snt    # Wrd   |  Corr        Sub       Del       Ins       Err     S.Err   |
|  Sum/Avg  |   130      2565   |  92.1        3.5       4.3       0.4       8.2      51.5   |
```
But if i use 3 gpu and increase the size of batch 3 times means 30*3=90, for reducing the total execution time 1/3, then performance is getting very very poor.

```
# CER results in exp/train_nodev_pytorch_train_mtlalpha1.0/decode_test_decode_ctcweight1.0_lm_word100/result.txt
|  SPKR     |  # Snt    # Wrd   |  Corr        Sub       Del       Ins       Err     S.Err   |
|  Sum/Avg  |   130      2565   |  25.0       23.8      51.2       1.9      76.9     100.0   |
```
Are you guys facing same problem? Looking forward to any solution. Thanks in advance.

**Comment by lazzylee** [comment]
you need decrease batchsize 30 -> 10 to get similar scheduler/warmup  ?

**Comment by mahbubnoor** [comment]
If i decrease the batch size 30 to 10 and use 3 GPU then will the execution be faster or more slow?

**Comment by lazzylee** [comment]
speed is similar to 1 GPU.  or faster.  

to my experience.
if you use default noam optimizer, you increase the effective batchsize to 9x (3*3) , it will affect your learning rate scheduler 
if your warmup iteration is too big, you may  always in warmup till the end.  you may not reach the peak lrate

**Comment by mahbubnoor** [comment]
So, if i want to finish it quickly means the total execution time i would like to reduce 3x times by using 3 gpu instead of 1 gpu, what would you suggest me so that i will get same results what i got using 1 gpu?

**Comment by lazzylee** [comment]
to my experience, if batchsize is 16+, more gpu will faster. (3x -> 2.x x),   less  communication cost between GPU
set larger batchsize will reduce training time too
but you need to tune warmup iterations and  lrate.
for example:          warmup_steps  to   1/9 ?

**Comment by mahbubnoor** [comment]
Thank you, But where can i set this warmup step parameter? I don't see this parameter settings in an4 recipe. And what is that 1/9? Another thing is that, your last response is opposite to your 1st response, In your 1st response you told to decrease the batch size 30 to 10. But here in the last you are telling to increase the batch size. I am actually confused. Actually in the given recipe it was already 30 batch size, So running it quickly using 3 gpu i increased the batch size 3 times means i used 90 batch size when i used 3 gpu. So execution finished quickly but result was poor which i already mentioned in my post.

**Comment by sw005320** [comment]
As @lazzylee discussed, the (virtual) batchsize is one of the optimization hyperparameters.
By controlling the batch size, we can mimic the single GPU result with the multiple GPUs, but it would not have speed benefits for small data like `an4`.
I recommend @mahbubnoor to apply it to the other larger corpus with the multiple GPUs.
I think you'd have multiple GPU benefits then.
